File: maps/src/main/java/it/localhostsoftware/maps/model/bitmapDescriptorFactory/BitmapDescriptorFactory.java
package it.localhostsoftware.maps.model.bitmapDescriptorFactory;

import android.content.Context;
import android.graphics.Bitmap;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GoogleApiAvailability;

import it.localhostsoftware.maps.model.bitmapDescriptor.BitmapDescriptor;

public interface BitmapDescriptorFactory {
    float HUE_RED = 0.0F;
    float HUE_ORANGE = 30.0F;
    float HUE_YELLOW = 60.0F;
    float HUE_GREEN = 120.0F;
    float HUE_CYAN = 180.0F;
    float HUE_AZURE = 210.0F;
    float HUE_BLUE = 240.0F;
    float HUE_VIOLET = 270.0F;
    float HUE_MAGENTA = 300.0F;
    float HUE_ROSE = 330.0F;

    static BitmapDescriptorFactory getInstance(Context context) {
        if (GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(context) == ConnectionResult.SUCCESS)
            return new GoogleBitmapDescriptorFactory();
        else return null;
    }

    BitmapDescriptor fromResource(int var0);

    BitmapDescriptor fromAsset(String var0);

    BitmapDescriptor fromFile(String var0);

    BitmapDescriptor fromPath(String var0);

    BitmapDescriptor defaultMarker();

    BitmapDescriptor defaultMarker(float var0);

    BitmapDescriptor fromBitmap(Bitmap var0);
}

File: maps/src/main/java/it/localhostsoftware/maps/model/BitmapDescriptorFactory.kt
package it.localhostsoftware.maps.model

import android.content.Context
import android.graphics.Bitmap
import it.localhostsoftware.maps.MobileServices
import it.localhostsoftware.maps.getMobileServices
import it.localhostsoftware.maps.google.model.GoogleBitmapDescriptorFactory
import it.localhostsoftware.maps.huawei.model.HuaweiBitmapDescriptorFactory

interface BitmapDescriptorFactory<BD : BitmapDescriptor<*>> {
    companion object {
        fun getInstance(c: Context) =
            when (c.getMobileServices()) {
                MobileServices.GOOGLE -> GoogleBitmapDescriptorFactory()
                MobileServices.HUAWEI -> HuaweiBitmapDescriptorFactory()
                else -> throw IllegalStateException()
            }

        const val HUE_RED = 0.0f
        const val HUE_ORANGE = 30.0f
        const val HUE_YELLOW = 60.0f
        const val HUE_GREEN = 120.0f
        const val HUE_CYAN = 180.0f
        const val HUE_AZURE = 210.0f
        const val HUE_BLUE = 240.0f
        const val HUE_VIOLET = 270.0f
        const val HUE_MAGENTA = 300.0f
        const val HUE_ROSE = 330.0f
    }

    fun fromResource(var0: Int): BD
    fun fromAsset(var0: String): BD
    fun fromFile(var0: String): BD
    fun fromPath(var0: String): BD
    fun defaultMarker(): BD
    fun defaultMarker(var0: Float): BD
    fun fromBitmap(var0: Bitmap): BD
}